### ISSUE TITLE
Tables Custom rendering

### ISSUE THREAD

**Issue by neildchen** [bootstrap-vue/bootstrap-vue#429]
I am a totally newbie here.

In the Tables Custom rendering example, it says:
```
<template slot="name" scope="field">
      {{field.name.first}} {{field.name.last}}
</template> 
``` 

So, why in the example after "Custom rendering" section:
```
<template slot="actions" scope="item">
       <b-btn size="sm" @click="details(item.item)">Details</b-btn>
</template>

```
shouldn't the scope be "field" ?

And in fact, I have tried both (scope="item" and scope="field" ), but the button won't show up in my page.... I am totally lost... (but the custom rendering button seems to work on jsfiddle no problem....)

Here is snippet of my code:
```
<template>
<div>
<b-table striped bordered hover show-empty responsive :items="items" :fields="fields">
      <template slot="wonumber" scope="item">
        {{item.wonumber}}
      </template>
      <template slot="status" scope="item">
        {{item.value.status}}
      </template>
      <template slot="actions" scope="field">
        <b-btn variant="primary" @click="details(item.item)">Change Status</b-btn>
      </template>-->
    </b-table>
</div>
</template>
<script>
export default {
  data() {
    return {
      items: [
          {
          wonumber: 'wonumber_001',
          status: true
        }, {
          wonumber: 'wonumber_002',
          status: true
        }
      ],
      fields: {
        wonumber: {
          label: 'WO Number'
        },
        status: {
          label: 'Status'
        },
        actions: {
          label: 'Actions'
        }
      }
    }
  },
  methods: {
    details(item) {
      alert(JSON.stringify(item))
    }
  }
}
</script>
```

I am using the newest 0.16.1, everything else seem working properly, except the custom button won't show up... Thank you in advance for any tips.

**Comment by tmorehouse** [comment]
Oops, yeah. Scope is "field"So, in your "fields" object, make sure you have the virtual column defined (for the button field). And then create a template to fill that slot, using the field's key as the slot name.                                                                                                                                                                                                                                                                                                                                         Sent from my BlackBerry 10 smartphone on the Bell network.                                                                                                                                                                                                                From: neilSent: Saturday, May 20, 2017 12:52 AMTo: bootstrap-vue/bootstrap-vueReply To: bootstrap-vue/bootstrap-vueCc: SubscribedSubject: [bootstrap-vue/bootstrap-vue] Tables Custom rendering (#429)I am a totally newbie here.
In the Tables Custom rendering example, it says:
<template slot="name" scope="field">
      {{field.name.first}} {{field.name.last}}
</template> 

So, why in the example after "Custom rendering" section:
<template slot="actions" scope="item">
       <b-btn size="sm" @click="details(item.item)">Details</b-btn>
</template>


shouldn't the scope be "field" ?
And in fact, I have tried both (scope="item" and scope="field" ), but the button won't show up in my page.... I am totally lost... (but the custom rendering button seems to work on jsfiddle no problem....)
Here is snippet of my code:
<template>
<div>
<b-table striped bordered hover show-empty responsive :items="items" :fields="fields">
      <template slot="wonumber" scope="item">
        {{item.wonumber}}
      </template>
      <template slot="status" scope="item">
        {{item.value.status}}
      </template>
      <template slot="actions" scope="field">
        <b-btn variant="primary" @click="details(item.item)">Change Status</b-btn>
      </template>-->
    </b-table>
</div>
</template>
<script>
export default {
  data() {
    return {
      items: [
          {
          wonumber: 'wonumber_001',
          status: true
        }, {
          wonumber: 'wonumber_002',
          status: true
        }
      ],
      fields: {
        wonumber: {
          label: 'WO Number'
        },
        status: {
          label: 'Status'
        },
        actions: {
          label: 'Actions'
        }
      }
    }
  },
  methods: {
    details(item) {
      alert(JSON.stringify(item))
    }
  }
}
</script>

I am using the newest 0.16.1, everything else seem working properly, except the custom button won't show up... Thank you in advance for any tips.

—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/bootstrap-vue/bootstrap-vue","title":"bootstrap-vue/bootstrap-vue","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/bootstrap-vue/bootstrap-vue"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"Tables Custom rendering (#429)"}],"action":{"name":"View Issue","url":"https://github.com/bootstrap-vue/bootstrap-vue/issues/429"}}}

**Comment by tmorehouse** [comment]
Take a look at this fiddle https://jsfiddle.net/pi0/a5s95a3q/

The actions slot template is `sope="item"` (to be rendered for each item)

**Comment by alexsasharegan** [comment]
Is this the goal end result? https://jsfiddle.net/a5s95a3q/33/

**Comment by alexsasharegan** [comment]
And does that mean the doc explanation for custom rendering is off? 

![image](https://cloud.githubusercontent.com/assets/16269252/26277412/ad57b066-3d3b-11e7-8076-2483116f683d.png)

**Comment by alexsasharegan** [comment]
Should it read like so:

### Custom rendering
Custom rendering for each field is possible using **scoped slots**.
If you want to add an extra field which does not exits on records, just add it to `fields` object.  Example:

```js
{
    fields: {
        name: {
            label: 'Person Full name',
            sortable: true
        },
    }
}
```

```html
<template slot="name" scope="item">
      {{item.value.first}} {{item.value.last}}
</template>
```

**Comment by tmorehouse** [comment]
Check out PR #433 

I've added some clarity and improved example for scope slots.  I know the documentation (and scoped slots)  confused me a bit, so I've updated it to make usag more clear.

**Comment by alexsasharegan** [comment]
You beat me to it! My additions are nowhere near as comprehensive though (I haven't gotten to make use of the table component yet).

Nice work! 👍

**Comment by tmorehouse** [comment]
Yeah, the user of the word `field` and/or `item` (in the scope variable) was confusing to me, and after reviewing the code and scoped slots docs it started to become more clear to me.

**Comment by alexsasharegan** [comment]
Template tags and scoped slots are something I'm just starting to dive into. Haven't needed them until recently.

**Comment by neildchen** [comment]
@alexsasharegan  thank you, that (https://jsfiddle.net/a5s95a3q/34/) is what I wanted but I couldn't have it worked. I will try your example again later when I have some time. Many thanks to you.
@tmorehouse thank you for the explanation. I will take a look again at the updated documentation. Thank you.

**Comment by tmorehouse** [comment]
@neildchen How is the custom field rendering working for you now?

**Comment by alexsasharegan** [comment]
Haven't gotten the chance to work on an electron project yet. Happy hunting on that one!

**Comment by neildchen** [comment]
@alexsasharegan Thank you Alex. I think you can give electron a try. It is quite fun playing with. Cheers.

**Comment by juannenna** [comment]
hi, maybe its not the place, but is there a way to use the same item.value in two different slots?? in my table i need two separate buttons, each of which triggering an event with the id of the item rendered in the row. how can i do that? thanks!

**Comment by tmorehouse** [comment]
@juannenna Yep, using a "virtual" field, and a custom slot for that field.

Check out https://github.com/bootstrap-vue/bootstrap-vue/blob/master/docs/components/table/README.md for the latest readme docs on table usage

**Comment by juannenna** [comment]
perfect! thanks, i wasn't aware of the "data" scope. i was only using/trying to use "fields" and "item".

Thanks for the quick response!

**Comment by tmorehouse** [comment]
`field` and `item` are just names you assign to the local scope variable, but can be confusing if you are thinking of the `fields` and `items` props.  using `data` or some other name for the variable makes it a bit ore clear as to it's use.

**Comment by alexsasharegan** [comment]
@neildchen & @juannenna If you're enjoying bootstrap-vue, let us know with a ⭐️ ! We appreciate your feedback!

**Comment by neildchen** [comment]
Hello @alexsasharegan I already have. Thank you.

**Comment by juannenna** [comment]
@alexsasharegan done! thanks!
